File: 2/sol.clj
(require '[clojure.string :as string])

(defn get-lines [file]
  (with-open [rdr (clojure.java.io/reader file)]
    (doall (line-seq rdr))))

(defn parse-line [line]
  (let [[d i] (string/split line #" ")]
    [d (Integer/parseInt i)]))

(def lines
  (map parse-line (get-lines "./input.txt")))

(defn step-part1 [x y cmd i]
  (cond
    (= cmd "forward") [(+ x i) y]
    (= cmd "up") [x (- y i)]
    :else [x (+ y i)]))

(defn step-part2 [x y aim cmd i]
  (cond
    (= cmd "forward") [(+ x i) (- y (* aim i)) aim]
    (= cmd "up") [x y (+ aim i)]
    :else [x y (- aim i)]))

;; list of commands, init vector, step function to reduce
(defn solve [cmds init step-fn]
  (reduce #(apply step-fn (into %1 %2)) init cmds))

(apply * (solve lines [0 0] step-part1))
(apply * (take 2 (solve lines [0 0 0] step-part2)))

File: 10/sol.py
with open("input.txt", "r") as f:
    txt = f.read()    
    lines = list(txt.splitlines())      

m = {"{":"}", "(":")", "[":"]", "<":">"}

def parse(l):
    s = []
    for c in l:
        if c in "{(<[":    
            s.append(c)    
        elif s and m[s[-1]] == c:
            s.pop()
        else:
            return s, c
    return s, None

def part1(lines):
    s = 0
    mm = {")":3, "]":57, "}":1197, ">":25137}
    for l in lines:
        _, i = parse(l)    
        if i is None:  
            continue    
        else:
            s += mm[i]    
    return s

def part2(lines):
    sl = []
    mm = {"(":1, "[":2, "{":3, "<":4}
    for l in lines:
        s, i = parse(l)
        if i is not None:    
            continue    
        rr = 0
        for c in reversed(s): 
            rr *= 5
            rr += mm[c]
        sl.append(rr)
    return sorted(sl)[len(sl) >> 1]

print(part1(lines))
print(part2(lines))
